File: twoSum/twosum01v2.js
// question: https://www.algoexpert.io/questions/two-number-sum
//PREP
// P: an array of numbers, a target number
// R: returns the two numbers that add to target number, if no two numbers are found return an empty array
// E: twoNumberSum([1,])
// P: initlaize a set
// loop through array put number in set when its found
// attempt to reach target number, if not reached then continue adding number to set
// if target number found return the two numbers.
// else return an empty array
function twoNumberSum(array, targetSum) {
  let numberSet = new Set();

  for (let idx = 0; idx < array.length; idx++) {
    let numberToLookFor = targetSum - array[idx];
    if (numberSet.has(numberToLookFor)) {
      return [numberToLookFor, array[idx]];
    } else {
      numberSet.add(array[idx]);
    }
  }

  return [];
}

File: twoSum/twoSum01v2.cs
using System;
using System.Collections.Generic;

public class Program {
  public static int[] TwoNumberSum(int[] array, int targetSum) {
      var numberToLookFor = 0;

      HashSet<int> hash = new HashSet<int>();

      for(var i = 0; i < array.Length; i++){
          int value = 0;
          numberToLookFor = targetSum - array[i];
          
          if( hash.TryGetValue(numberToLookFor, out value) ){
              int[] ans = { numberToLookFor , array[i] };
              return ans;
          } else {
              hash.Add(array[i]);
          }
      }

    return new int[0];
  }
}

